### PR TITLE
blog: Add GSoC'24 blog on debugging mimalloc port

### PR DESCRIPTION
**Pull request by huyang531**
Signed-off-by: Yang Hu [yanghuu531@gmail.com](mailto:yanghuu531@gmail.com)

### CODE REVIEW

**Comment by StefanJum** [comment]
@huyang531 I see this is marked as draft. Any reason for that?

**Comment by huyang531** [comment]
> @huyang531 I see this is marked as draft. Any reason for that?

Not really. I just marked this ready for review.

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 1]
```diff
@@ -0,0 +1,158 @@
+---
+title: "GSoC'24: Debugging the Multithreaded Memory Allocation Interface"
+description: |
+  Different applications have different memory usage patterns and perform better
+  when using dedicated memory allocators. Right now, the buddy allocator is the only
+  available memory allocator used in Unikraft, so this GSoC project aims to 
+  port more memory allocators to it, starting with mimalloc.
+publishedDate: 2024-07-12
```
Comment (StefanJum) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:8]: Change the date

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+The code for this logic though, is customized for the buddy allocator:
+
+1. Initialize a physically-mapped small heap size of 16 pages
+1. Initialize the allocator based on this heap
+1. Initialize the VAS
+1. Map the rest of the heap as an anonymous VMA
+1. Call `uk_alloc_addmem()` to let the buddy allocator know that this anonymous region that is at its disposal
+
+This design accommodates the fact that the buddy allocator needs initial memory space to store its metadata and that it needs information about the total available memory for initializing its free lists.
+However, for *mimalloc*, we are not initializing the allocator per-se at this time because the Thread Local Storage (TLS) infrastructure that it needs are not yet ready.
```
Comment (StefanJum) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:80]:
```suggestion
However, for *mimalloc*, we are not initializing the allocator per-se at this time because the Thread Local Storage (TLS) infrastructure that it needs is not yet ready.
```

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+[   11.510388] CRIT: <idle> {r:0x107240,f:0x28cf00} [libkvmplat] <trace.c @   86> base is 0x11a70 caller is 0x1732e6
+```
+
+After extensive debugging, I realized that this bug is not unique to the *mimalloc* port.
+I reproduced a similar error with the same library configurations on the buddy allocator as well and I have submitted [an issue](https://github.com/unikraft/unikraft/issues/1475) addressing this.
+
+The current `ukvmem` implementation is also incomplete as some of its key data structures are not CPU-local.
+For example,
+
+At `vmem.c:27`:
```
Comment (StefanJum) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:114]: Link the file on github.

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+```c
+/*
+ * Pointer to currently active virtual address space.
+ * TODO: This should move to a CPU-local variable
+ */
+static struct uk_vas *vmem_active_vas;
+```
+
+And `paging.c:97`:
```
Comment (StefanJum) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:124]: Link the file

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thanks @huyang531, see the inlines.

Also @razvanvirtan  @RaduNichita please do a content check.

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx (hunk 3 above)
Comment (huyang531) [reply content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx]: @mogasergiu has pointed out that this should not be a concern at this time so this part has been removed

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+- allocators
+- synchronization
+---
+
+## Project Overview
+
+### Memory allocation in Unikraft
+
+Memory allocators have a significant impact on application performance.
+[Research](https://dl.acm.org/doi/10.1145/378795.378821) have shown that programs can run as much as 60% faster just by switching to an appropriate memory allocator.
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:24]: There have been some [research papers](https://dl.acm.org/doi/10.1145/378795.378821) which compared different memory allocators. According to their work, Switching to the appropiate memory allocator may improve the application performance by 60%.

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+Unikraft currently supports only one general-purpose memory allocator, the buddy allocator.
+This project aims to port [mimalloc](https://github.com/microsoft/mimalloc) (pronounced "me-malloc"), a high-performance general-purpose memory allocator developed by Microsoft, to Unikraft.
+This is the first step in a series of efforts to provide Unikraft users with more memory allocators to choose from to optimize the performance of their applications.
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:27]:
```suggestion
This is the first step in a series of efforts to provide Unikraft users with more memory allocators to choose from to enhance the performance of their applications.
```

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+### Objectives
+
+[Hugo Lefeuvre](https://github.com/hlef) has made an effort to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:31]:
```suggestion
[Hugo Lefeuvre](https://github.com/hlef) made an initial work to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
```

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+However, as Unikraft has evolved significantly over the years, more work is needed to adapt the `lib-mimalloc` port to the latest Unikraft core.
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:32]:
```suggestion
However, as Unikraft has evolved significantly over the years, more work has to be done in order to adapt the `lib-mimalloc` port to the latest Unikraft core.
```

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+So far, I have successfully ported the mimalloc memory allocator to Unikraft, marked by a successful compilation of mimalloc `v1.6.3` against the latest Unikraft core (v0.17.0) with `lib-musl`.
+Now, I am focused on validating the port and testing its performace.
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:35]:
```suggestion
Now, I am focused on validating the functionality and benchmarking the performance of the allocator.
```

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+After that, the goal is to port the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0).
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:36]:
```suggestion
The next step would be to port the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0).
```

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+## Current Progress
+
+Following porting the multithreaded `cache-scratch` stress test in [the last blog post](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft), this past three weeks I have been debugging the it.
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:40]:
```suggestion
Following porting the multithreaded `cache-scratch` stress test in [the last blog post](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft), this past three weeks I have been debugging the it.
```

This is unclear to me, can you rephrase?

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+### Debugging `pthread`
+
+The [first error we ran into](https://unikraft.org/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft#testing) was a memory corruption error with the `pthread` interface.
+After investigating the `pthread_join()` calls, I found that the threads were never even created correctly at all because `pthread_create()` uses the POSIX `mmap` system call, which never existed as we did not turn on virtual memory support (i.e., `ukvmem` alongside `posix-mmap`).
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:45]: I don't think you should go that deep into technical details

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+After extensive debugging, I realized that this bug is not unique to the *mimalloc* port.
+I reproduced a similar error with the same library configurations on the buddy allocator as well and I have submitted [an issue](https://github.com/unikraft/unikraft/issues/1475) addressing this.
+More work needs to be done to fix the virtual memory infrastructure.
+
+## Next steps
+
+In order to resolve the abovementioned issues, I still have much to study about Unikraft's virtual memory management subsystem, its boot process, and how *mimalloc* works under the hood - and I will focus on just that in the following weeks.
+
+Hopefully, with those issues fixed, we could finally get the `cache-scratch` test running on multiple cores.
+Once that is done, I will get back on track with my "global" TODO list:
```
Comment (RaduNichita) [inline content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:118]: Can you mention what ideas do you have for benchmarking?

**Review by RaduNichita — CHANGES_REQUESTED** [review]

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx (hunk 13 above)
Comment (huyang531) [reply content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:118]: I have written much about benchmarking (i.e., which benchmarks I would use, what they are for, etc.) in my last blog post so I didn't elaborate here. I will attach a link to my last blog post

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx (hunk 11 above)
Comment (huyang531) [reply content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:40]: Done

File: content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx (hunk 12 above)
Comment (huyang531) [reply content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx:45]: I wrote the blog in sort of chronological order - it was because of this`pthread` bug that I turned on virtual memory support, hence the rest. I have shortened the paragraph but still kept it here.

**Review by StefanJum — APPROVED** [review]
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>
